core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala

```diff
@@ -272,17 +272,41 @@ class ExecutorAllocationManagerSuite
   test("ignore task end events from completed stages") {
     sc = createSparkContext(0, 10, 0)
     val manager = sc.executorAllocationManager.get
+
+    // We simulate having a stage fail, but with tasks still running.  Then another attempt for
+    // that stage is started, and we get task completions from the first stage attempt.  Make sure
+    // the value of `totalTasksRunning` is consistent as tasks finish from both attempts (we only
+    // count tasks from the active attempt)
     val stage = createStageInfo(0, 5)
     post(sc.listenerBus, SparkListenerStageSubmitted(stage))
     val taskInfo1 = createTaskInfo(0, 0, "executor-1")
     val taskInfo2 = createTaskInfo(1, 1, "executor-1")
     post(sc.listenerBus, SparkListenerTaskStart(0, 0, taskInfo1))
     post(sc.listenerBus, SparkListenerTaskStart(0, 0, taskInfo2))
 
+    // the first attempt finishes (eg. with a fetch failure).  The tasks haven't completed yet, but
+    // we ignore the running tasks.
```

Comment:
again, update comment "The tasks in the zombie attempt haven't completed yet, so we still count them"